chorus_lib/src/core.rs

```diff
@@ -89,6 +89,85 @@ where
     }
 }
 
+// --- HList and Helpers ---
+
+/// heterogeneous list
+pub trait HList {
+    /// returns
+    fn to_string_list() -> Vec<&'static str>;
+}
+/// end of HList
+pub struct HNil;
+/// An element of HList
+pub struct HCons<Head, Tail>(Head, Tail);
+
+impl HList for HNil {
+    fn to_string_list() -> Vec<&'static str> {
+        Vec::new()
+    }
+}
+impl<Head, Tail> HList for HCons<Head, Tail>
+where
+    Head: ChoreographyLocation,
+    Tail: HList,
+{
+    fn to_string_list() -> Vec<&'static str> {
+        let mut v = Tail::to_string_list();
+        v.push(Head::name());
+        v
+    }
+}
+
+// TODO(shumbo): Export the macro under the `core` module
+
+/// Macro to generate hlist
+#[macro_export]
+macro_rules! hlist {
```

Comment:
1. needs a more explicit name (`LocationSet`?)
2. Because this macro expands to a type, it should be in upper camel case, as per this [issue](https://github.com/rust-lang/api-guidelines/issues/145)